contracts/cw1-subkeys/src/contract.rs

```diff
@@ -243,19 +354,27 @@ fn can_send<S: Storage, A: Api, Q: Querier>(
     if cfg.is_admin(&owner_raw) {
         return Ok(true);
     }
-
-    // we only accept bank messages - ensure type and get the amount
-    let amount = match msg {
-        CosmosMsg::Bank(BankMsg::Send { amount, .. }) => amount,
-        _ => return Ok(false),
-    };
-
-    // now we check if there is enough allowance for this message
-    let allowance = allowances_read(&deps.storage).may_load(owner_raw.as_slice())?;
-    match allowance {
-        // if there is an allowance, we subtract the requested amount to ensure it is covered (error on underflow)
-        Some(allow) => Ok(allow.balance.sub(amount).is_ok()),
-        None => Ok(false),
+    match msg {
+        CosmosMsg::Bank(BankMsg::Send { amount, .. }) => {
+            // now we check if there is enough allowance for this message
+            let allowance = allowances_read(&deps.storage).may_load(owner_raw.as_slice())?;
+            match allowance {
+                // if there is an allowance, we subtract the requested amount to ensure it is covered (error on underflow)
+                Some(allow) => Ok(allow.balance.sub(amount).is_ok()),
+                None => Ok(false),
+            }
+        }
+        CosmosMsg::Staking(staking_msg) => {
+            permissions_read(&deps.storage)
```

Comment:
Interesting chain. 

I think the `?` and `match` approach for allowances is more legible. But nice that you learned a bit more about map and map_or and handling Options and Results. Key skills for working with rust

Reply:
Just some old scala and functional programming skills λ

Reply:
Applied the `?` and `match` case.
But `can_send` method looks weird to me. The return value is `StdResult` but in tests, it is expected to be always boolean. Which throws the error information to the thrash 🤔

Reply:
Error should only throw if we have trouble parsing from disk, and we cannot *prove* that will never happen.

In all normal cases (non-corrupt db), we should get true or false. You would have to write invalid json to the key to trigger an error